airbyte-integrations/bases/base-typing-deduping/src/main/java/io/airbyte/integrations/base/destination/typing_deduping/TypeAndDedupeOperationValve.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.integrations.base.destination.typing_deduping;
+
+import io.airbyte.protocol.models.v0.AirbyteStreamNameNamespacePair;
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Supplier;
+
+/**
+ * A slightly more complicated way to keep track of when to perform type and dedupe operations per
+ * stream
+ */
+public class TypeAndDedupeOperationValve extends ConcurrentHashMap<AirbyteStreamNameNamespacePair, Long> {
+
+  private static final long TWO_MINUTES_MILLIS = 1000 * 60 * 2;
+
+  private static final long FIVE_MINUTES_MILLIS = 1000 * 60 * 5;
+
+  private static final long TEN_MINUTES_MILLIS = 1000 * 60 * 10;
+
+  private static final long FIFTEEN_MINUTES_MILLIS = 1000 * 60 * 15;
```

Comment:
Maybe add a comment that 15min is the max time between checkpoints defined by the protocol.

Reply:
done